### PR TITLE
Abstract launch trait

### LINKED ISSUE #7387: Customize `NodeBuilder::launch`
### Describe the feature

Abstract the launch method so `NodeBuilder` can have a different launch sequence.

I would like to specify which components are launched by my node an how they are launched. This allows me to implement a different consensus engine.

### Additional context

This should be a step towards custom consensus for the full node.

### PR DESCRIPTION
**Pull request by grantkee**
- `LaunchNode` trait to abstract `NodeBuilder::launch`
- impl `LaunchNode` for `DefaultLauncher` to replace the original launch method for node builder
- update examples
- make launch types publicly available for other to implement `launch()`
    - rpc hooks
    - node hooks
    - exex
    - builder context
- implement default for `RpcHooks` for clippy lint
- update docs 

closes #7387

### CODE REVIEW

**Comment by grantkee** [comment]
The current PR doesn't really solve the issue. `NodeBuilder` still has the same launch method. Ideally, anyone can still use NodeBuilder but set a custom launch.

One idea I have is to take a similar approach as Hooks and include a generic on `NodeBuilder` for `LaunchNode`, but that isn't a good approach I think because of all the associated types with `FullNodeComponents`.

**Review by mattsse — CHANGES_REQUESTED** [review]
I experimented with this a bit, I think we can encapsulate this very easily with a trait:

```rust 
trait LaunchNode<...>

async fn launch(self, ....) -> eyre::Result<NodeHandle<...>>
```

then we add:
```rust
async fn launch_with(self, launcher: L) where L: LaunchNode<....>
```

and a

```rust
struct DefaultLauncher;
```

that contains everything we have right now in the launch fn, so that launch becomes:

```rust
async fn launch(self) {
   self.lanch_with(DefaultLauncher::default()).await 
}
```

This will also become useful to split dev launch vs regular launch mode, so very supportive here

**Comment by grantkee** [comment]
I'll take this on. Thanks for the guidance @mattsse !

File: crates/node-builder/src/builder.rs
Comment (grantkee) [inline crates/node-builder/src/builder.rs:69]: @mattsse should I update the rest of builder.rs to use these types? We explicitly use them in a lot of method signatures. I just want to confirm it wouldn't pigeon-hole us down the road

Comment (grantkee) [reply crates/node-builder/src/builder.rs:69]: I'd also like to know your preference on consolidating `N` and `Types` for these types. I personally found `Types` to be clearer for understanding, but `N` is more inline with rust conventions. I can also rename `Components` to just `C` if you prefer

Comment (grantkee) [reply crates/node-builder/src/builder.rs:69]: Maybe it's best to leave it as a single abbreviation `N` in the type definition and use `Types` in method signatures for clarity. Your call

Comment (mattsse) [reply crates/node-builder/src/builder.rs]: > Maybe it's best to leave it as a single abbreviation `N` in the type definition and use `Types` in method signatures for clarity. Your call

that makes sense, but I want to do this seperately

File: crates/node-builder/src/builder.rs
Comment (mattsse) [inline crates/node-builder/src/builder.rs:163]: if we make these pub then the launch trait is kinda pointless?

Comment (grantkee) [reply crates/node-builder/src/builder.rs:163]: I don't think so because devs need access to these fields within the launch trait. For example, the second line of code in the default launch trait impl deconstructs `NodeBuilder` then uses the database to create the node's `ProviderFactory`.

If we make these fieldss private, then that forces devs to impl their own node builder struct and they lose access to a lot of great methods/features `NodeBuilder` provides

Comment (grantkee) [reply crates/node-builder/src/builder.rs]: @mattsse I restored the fields to private and added `WithLaunchContext::launch_with`

File: crates/node-builder/src/rpc.rs
Comment (grantkee) [inline crates/node-builder/src/rpc.rs:43]: I will follow up in a separate PR to expose calling hooks through a public API:
```
pub fn call_on_rpc_started(&self, ctx, handles) -> eyre::Result<()> {
  self.on_rpc_started.on_rpc_started(ctx, handles)
}
```

**Review by mattsse — COMMENTED** [review]
I like this direction, will alloc some time to get this in tmrw!

ty

**Comment by grantkee** [comment]
@mattsse I'm excited to hear you like this direction! I'd like to continue expanding reth as an agnostic EVM execution client with your guidance.

Is there anything else I can do to help get this PR merged?

File: crates/node-builder/src/components/traits.rs
Comment (mattsse) [inline crates/node-builder/src/components/traits.rs:185]: can we put this on a container type, this way we can make changes to this more easily

File: crates/node-builder/src/components/traits.rs
Comment (mattsse) [inline crates/node-builder/src/components/traits.rs:160]: can we add a blanket impl for Fn that returns a future, similar to:

https://github.com/paradigmxyz/reth/blob/d6dd1c2dabbc1801c1b68da357bf7fcaff4a47a0/crates/node-builder/src/components/payload.rs#L20-L26

File: crates/node-builder/src/builder.rs
Comment (mattsse) [inline crates/node-builder/src/builder.rs:467]: hmm, why is is this necessary now?

**Review by mattsse — CHANGES_REQUESTED** [review]
sorry for the delay,

for some reason I'm unable to push to this pr...
could you please check that this pr allows edits by maintainer?

I have a few nits

**Comment by grantkee** [comment]
Sorry, @mattsse - I didn't realize edit access for maintainers is different for organization forks. I added you as a collaborator to the forked repo under layer-g org. I'll avoid creating forks through my organization moving forward for simplicity.

Here's more context if you're interested https://github.com/orgs/community/discussions/5634

File: crates/node-builder/src/builder.rs
Comment (grantkee) [reply crates/node-builder/src/builder.rs:467]: I'll take another look at this

Comment (mattsse) [reply crates/node-builder/src/builder.rs:467]: cool, I'm still very supportive of this feature just haven't entirely settled on a design yet, but I think the Launch trait is the best we can do atm

File: crates/node-builder/src/components/traits.rs
Comment (grantkee) [reply crates/node-builder/src/components/traits.rs:160]: great idea!

**Comment by mattsse** [comment]
thanks for exploring this @grantkee 

I refactored the node builder internals quite a bit,

there is now a Launch trait with #7847

and starting with #7884 we're pushing more setup logic to a helper type.

there are likely still a few limitations, like private types etc, so please take another look at main and submit issues for things you'd like to have

**Pull request closed without merging**
